### DISCUSSION TITLE
Is using the agent and the starter at the same time possible ?

### DISCUSSION THREAD

**Discussion by Mumeii** [open-telemetry/opentelemetry-java-instrumentation#12936] [Q&A] [answered]
Hello 👋 

Disclaimer : I've seen [this discussion](https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/12386) that have a title quite related to the current one.

But in the end, the answers made in it didn't really answered my existential dread about using both agent and starter at the same time 😋 ... so here it is : 

**Is using both at the same time is supported by one, the other, or both ?**

If it is, are they tricky things to know to avoid some potential clashes or troubles ?

**Why am I asking this question ? :** 

I have a Spring software that is currently monitored thanks the OT agent.

Doing it that way is great because #ZeroCode 😄 

But at the same time, this software is relying on Spring friendly tool, which itself is able to produce spans thanks Open Telemetry.

The trouble is that it's producing those span only if it finds a bean implementing `io.opentelemetry.api.trace.Tracer` interface in the Spring context : `@ConditionalOnBean( io.opentelemetry.api.trace.Tracer.class )`

And unfortunately, this tool creator don't have much extra time to spend creating an agent extension ...

So in the end if I want the best of both worlds, the most straightforward option would be , if it's possible, to use both technics at the same time :

- add and configure the OT spring boot starter to have my tool finally producing spans
- keep using the agent for pretty much everything else

Thanks in advance for your enlightenment on that matter ! 🙏 

> PS : no matter it's possible or not, maybe adding an extra entry regarding this matter in the documentation coud be useful.
> Around [there](https://opentelemetry.io/docs/zero-code/java/spring-boot-starter/) maybe ?

**Comment by laurit** [comment]
> The trouble is that it's producing those span only if it finds a bean implementing io.opentelemetry.api.trace.Tracer interface in the Spring context : @ConditionalOnBean( io.opentelemetry.api.trace.Tracer.class )

You don't need to use the spring starter and the agent together to solve this. Although our spring starter does not define such a bean (it did a long time ago), you can easily define it yourself with something like
```java
  @Bean
  Tracer tracer(OpenTelemetry openTelemetry) {
    return openTelemetry.getTracer("test");
  }
```
Similarly with the agent you'd include opentelemetry-api in your application and use
```java
  @Bean
  Tracer tracer() {
    return GlobalOpenTelemetry.get().getTracer("test");
  }
```

**Reply by Mumeii** [reply]
Hi @laurit 

Thanks for your reply !

I tried your advice today and it's working like a breeze

The only drawback is that I'd also like to setup a custom SpanProcessor to have spans produced by my "Spring friendly tool" being prefixed

Indeed, our OPS are managing a centralized collector system, and they asked us to have generated span & metrics being prefixed by each developer team's name 😢 

What I see at my application startup is that it's the agent that is fist making use of the SdkTracerProviderBuilder.build() method.

And looking at SdkTracerProvider class, it don't seems possible to have its SpanProcessor list being modified afterward ...

Do you see any way to solve this second problematic ?

Apparently, it seems possible [to apply an AutoConfigurationCustomizer](https://opentelemetry.io/docs/zero-code/java/spring-boot-starter/sdk-configuration/#programmatic-configuration) to introduce a custom span processor whenever using the starter ... 

But in the end, I fall again on my initial question : Can both starter and agent get along in that case ? 😋

**Reply by laurit** [reply, marked as answer]
> And looking at SdkTracerProvider class, it don't seems possible to have its SpanProcessor list being modified afterward ...

The sdk configured by the agent can be changed with extensions, see https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension There is a sample `SpanProcessor` there.

> Apparently, it seems possible [to apply an AutoConfigurationCustomizer](https://opentelemetry.io/docs/zero-code/java/spring-boot-starter/sdk-configuration/#programmatic-configuration) to introduce a custom span processor whenever using the starter

This is pretty much the same as with agent extension, the only difference is that the `SpanProcessor` can reside inside your application.

> But in the end, I fall again on my initial question : Can both starter and agent get along in that case ? 😋

They are not intended to be used together. They configure separate opentelemtry sdks and contain the same instrumentations that could result in duplicate telemetry.

**Reply by Mumeii** [reply]
@laurit : Ok, it's clearer now 😋

Many thanks for your invaluable replies, and merry Christmas 🎄 🎅 🎁  !
